core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -833,7 +833,17 @@ public List<ManifestFile> apply(TableMetadata base, Snapshot snapshot) {
         filterManager.filterManifests(
             SnapshotUtil.schemaFor(base, targetBranch()),
             snapshot != null ? snapshot.dataManifests(ops.io()) : null);
-    long minDataSequenceNumber =
+
+    long minNewFileSequenceNumber =
+        addedDataFiles().stream()
+            .filter(x -> x.dataSequenceNumber() != null && x.dataSequenceNumber() >= 0)
+            .mapToLong(ContentFile::dataSequenceNumber)
+            .reduce(
+                newDataFilesDataSequenceNumber != null
+                    ? newDataFilesDataSequenceNumber
+                    : base.nextSequenceNumber(),
+                Math::min);
```

Comment:
Do we actually need to iterate through the `addedDataFiles`? 

If I understood the issue correctly, the problem is that it's possible for a user to commit a rewrite operation and specify an older data sequence number, and the current logic would drop delete files which actually need to still be referenced in the new commit since it's not considering the specified data file sequence number.

So I *think* all we would need to do here is keep the existing logic for determining minDataSequenceNumber and then also min that with the `newDataFilesDataSequenceNumber` if it's not null

```
long minNewDataSequenceNumber = <Existing logic>
if (newDataFilesDataSequenceNumber != null) {
    minNewDataSequenceNumber = Math.min(minNewDataSequenceNumber, newDataFilesDataSequenceNumber);
} 
```

Reply:
I tried the above and the new unit test still passes, though let me know if you see a flaw in my reasoning. cc @rdblue @aokolnychyi @findepi

Reply:
Currently _I think_ ('m also not 100% sure) you are right as there isn't an API that would allow setting a specific data file to a SN smaller than `newDataFilesDataSequenceNumber`. But I think that's just a side affect of the current operations inheriting from `MergingSnapshotProducer`. A new inheritor of `MergingSnapshotProducer` could simply be using `minNewDataSequenceNumber` as a default and pass in files with explicit SNs on top.